File: ios/WeatherDivKit/Data/DocumentRepository.swift
import DivKit

/// Named load policies over DocumentLoading. Port of Android document/DocumentRepository.
final class DocumentRepository {
    private let loader: DocumentLoading
    init(loader: DocumentLoading) { self.loader = loader }

    func coldStartLocal(lang: String) -> DocumentBundle? {
        loader.loadCache(lang: lang) ?? loader.loadBundledSkeleton()
    }

    func fetch(lang: String, lat: String?, lon: String?, name: String?) async -> DocumentBundle? {
        try? await loader.load(lang: lang, lat: lat, lon: lon, name: name)
    }

    func fetchOrCache(lang: String, lat: String?, lon: String?, name: String?) async -> DocumentBundle? {
        if let fresh = try? await loader.load(lang: lang, lat: lat, lon: lon, name: name) { return fresh }
        return loader.loadCache(lang: lang)
    }

    func citySearchPatch(query: String, lang: String) async -> DivPatch? {
        await loader.loadCitySearch(query: query, lang: lang)
    }
}

File: android/src/main/java/com/example/weatherdivkit/document/DocumentRepository.kt
package com.example.weatherdivkit.document

import com.yandex.div2.DivData
import com.yandex.div2.DivPatch

/** Named load policies over [DocumentSource]. All methods block; call off the main thread. */
class DocumentRepository(private val source: DocumentSource) {
    /** Cold-start phase 1: cache for this lang, else bundled skeleton. Never null. */
    fun coldStartLocal(lang: String): Map<Screen, DivData> =
        source.loadFromCache(lang) ?: source.loadFromAssets()

    /** Network-only. Cold-start ph2, PTR, setCity. Null → caller keeps current. */
    fun fetch(lang: String, lat: String?, lon: String?, name: String?): Map<Screen, DivData>? =
        source.loadFromNetwork(lang, lat, lon, name)

    /** Network, else same-lang cache. setLang. Never falls to bundled asset. */
    fun fetchOrCache(lang: String, lat: String?, lon: String?, name: String?): Map<Screen, DivData>? =
        source.loadFromNetwork(lang, lat, lon, name) ?: source.loadFromCache(lang)

    fun citySearchPatch(query: String, lang: String): DivPatch? =
        source.loadCitySearch(query, lang)
}
